### PR TITLE
Add useful information to understand governance proposals

### PR DESCRIPTION
**Pull request by nategraf**
### Description

It's rather difficult to tell from the current `governance` cli commands whether a proposal is passing or not, and when it will transition between stages. This PR adds from extra information to try and address that.

### Other changes

Hide the approver commands because average users will never be able to use them

### Tested

I've been using these changes to deploy governance fixes to a local testnet and baklava

### Backwards compatibility

100%

### CODE REVIEW

File: packages/cli/src/commands/governance/executehotfix.ts [hunk 1]
```diff
@@ -27,14 +28,15 @@ export default class ExecuteHotfix extends BaseCommand {
...
+    this.kit.defaultAccount = account
 
     const jsonString = readFileSync(res.flags.jsonTransactions).toString()
     const jsonTransactions: ProposalTransactionJSON[] = JSON.parse(jsonString)
 
     const builder = new ProposalBuilder(this.kit)
     jsonTransactions.forEach((tx) => builder.addJsonTx(tx))
     const hotfix = await builder.build()
-    const saltBuff = Buffer.from(res.flags.salt, 'hex')
+    const saltBuff = Buffer.from(trimLeading0x(res.flags.salt), 'hex')
```
Comment (yorhodes) [inline packages/cli/src/commands/governance/executehotfix.ts]: We could have a `parseBuffer` which resolves this 0x handling, but I like the changes

File: packages/cli/src/commands/governance/hashhotfix.ts [hunk 2]
```diff
@@ -0,0 +1,46 @@
...
+    ...BaseCommand.flags,
+    jsonTransactions: flags.string({
+      required: true,
+      description: 'Path to json transactions of the hotfix',
+    }),
+    salt: flags.string({ required: true, description: 'Secret salt associated with hotfix' }),
+  }
+
+  static examples = [
+    'hashhotfix --jsonTransactions ./transactions.json --salt 0x614dccb5ac13cba47c2430bdee7829bb8c8f3603a8ace22e7680d317b39e3658',
```
Comment (yorhodes) [inline packages/cli/src/commands/governance/hashhotfix.ts:25]: nice!

File: packages/cli/src/commands/governance/view.ts [hunk 3]
```diff
@@ -27,9 +28,33 @@ export default class View extends BaseCommand {
...
+
+    printValueMapRecursive({
+      ...record,
+      requirements: {
+        participation: participationParams.baseline,
+        agreement: constitution.times(100).toString() + '%',
+      },
+      isApproved: await governance.isApproved(id),
+      isProposalPassing: await governance.isProposalPassing(id),
+      secondsUntilStages: await governance.timeUntilStages(id),
```
Comment (yorhodes) [inline packages/cli/src/commands/governance/view.ts:53]: +100

File: packages/cli/src/commands/governance/viewhotfix.ts [hunk 4]
```diff
@@ -10,14 +10,20 @@ export default class ViewHotfix extends BaseCommand {
   static flags = {
     ...BaseCommand.flags,
     hash: flags.string({ required: true, description: 'Hash of hotfix transactions' }),
-    notyet: flags.boolean({
-      description: 'Whether to list validators who have or have not yet whitelisted',
+    whitelisters: flags.boolean({
+      description: 'If set, diplays validtors that have whitelisted the hotfix.',
```
Comment (yorhodes) [inline packages/cli/src/commands/governance/viewhotfix.ts]: nit: `validators`

File: packages/cli/src/commands/governance/viewhotfix.ts [hunk 5]
```diff
@@ -10,14 +10,20 @@ export default class ViewHotfix extends BaseCommand {
...
+      exclusive: ['nonwhitelisters'],
+    }),
+    nonwhitelisters: flags.boolean({
+      description: 'If set, diplays validtors that have not whitelisted the hotfix.',
```
Comment (yorhodes) [inline packages/cli/src/commands/governance/viewhotfix.ts]: ditto

File: packages/cli/src/commands/governance/viewhotfix.ts [hunk 6]
```diff
@@ -30,22 +36,26 @@ export default class ViewHotfix extends BaseCommand {
...
+    const tally = await governance.hotfixWhitelistValidatorTally(hash)
+    const quorum = await governance.byzantineQuorumValidators()
+    printValueMap({
+      tally,
+      quorum,
+    })
+
+    if (res.flags.whitelisters || res.flags.nonwhitelisters) {
       const validators = await this.kit.contracts.getValidators()
       const accounts = await validators.currentValidatorAccountsSet()
```
Comment (yorhodes) [inline packages/cli/src/commands/governance/viewhotfix.ts:49]: is this vulnerable to the recent `signerToAccount` bug?

File: packages/cli/src/commands/governance/viewhotfix.ts [hunk 7]
```diff
@@ -30,22 +36,26 @@ export default class ViewHotfix extends BaseCommand {
...
-      const tally = await governance.hotfixWhitelistValidatorTally(hash)
-      const quorum = await governance.byzantineQuorumValidators()
-      printValueMap({
-        tally,
-        quorum,
-      })
 
+    const tally = await governance.hotfixWhitelistValidatorTally(hash)
+    const quorum = await governance.byzantineQuorumValidators()
+    printValueMap({
```
Comment (yorhodes) [inline packages/cli/src/commands/governance/viewhotfix.ts:42]: makes sense to have a consistent output format

File: packages/contractkit/src/governance/proposals.ts [hunk 8]
```diff
@@ -28,6 +41,12 @@ export interface ProposalTransactionJSON {
   value: string
 }
 
+/**
+ * Convert a compiled proposal to a human-readble JSON form using network information.
```
Comment (yorhodes) [inline packages/contractkit/src/governance/proposals.ts]: +1 thanks for documenting!
auto docs ftw

Comment (yorhodes) [reply packages/contractkit/src/governance/proposals.ts]: nit: `readable`

File: packages/contractkit/src/wrappers/Governance.ts [hunk 9]
```diff
@@ -220,10 +228,11 @@ export class GovernanceWrapper extends BaseWrapper<Governance> {
 
   async timeUntilStages(proposalID: BigNumber.Value) {
     const meta = await this.getProposalMetadata(proposalID)
+    const now = Math.round(new Date().getTime() / 1000)
     const durations = await this.stageDurations()
-    const referendum = meta.timestamp.plus(durations.Approval)
+    const referendum = meta.timestamp.plus(durations.Approval).minus(now)
     const execution = referendum.plus(durations.Referendum)
-    const expiration = referendum.plus(durations.Execution)
+    const expiration = execution.plus(durations.Execution)
```
Comment (yorhodes) [inline packages/contractkit/src/wrappers/Governance.ts:307]: nice catch

File: packages/contractkit/src/wrappers/Governance.ts [hunk 10]
```diff
@@ -263,6 +272,30 @@ export class GovernanceWrapper extends BaseWrapper<Governance> {
     }
   }
 
+  /**
+   * Returns the required ratio of yes:no votes needed to exceed in order to pass the proposal.
+   * @param tx Transaction to determine the constitution for running.
+   */
+  async getConstitution(tx: ProposalTransaction): Promise<BigNumber.Value> {
```
Comment (yorhodes) [inline packages/contractkit/src/wrappers/Governance.ts]: can we make a utility for a tx array/`Proposal` type like the CLI uses?
@mcortesi is a big advocate of a "dumb" CLI and a "featureful" kit

File: packages/contractkit/src/wrappers/Validators.ts [hunk 11]
```diff
@@ -515,7 +515,11 @@ export class ValidatorsWrapper extends BaseWrapper<Validators> {
    */
   async currentValidatorAccountsSet() {
     const signerAddresses = await this.currentSignerSet()
-    const accountAddresses = await concurrentMap(5, signerAddresses, this.validatorSignerToAccount)
+    const accountAddresses = await concurrentMap(
+      5,
+      signerAddresses,
+      this.validatorSignerToAccount.bind(this)
```
Comment (yorhodes) [inline packages/contractkit/src/wrappers/Validators.ts]: what does this do?

**Review by yorhodes — APPROVED** [review]
Few spelling mistakes and suggestions

File: packages/cli/src/commands/governance/executehotfix.ts (hunk 1 above)
Comment (nategraf) [reply packages/cli/src/commands/governance/executehotfix.ts]: Switched to using `hexToBuffer`

File: packages/contractkit/src/wrappers/Validators.ts (hunk 11 above)
Comment (nategraf) [reply packages/contractkit/src/wrappers/Validators.ts]: Changed it to be more readable

File: packages/contractkit/src/wrappers/Governance.ts (hunk 10 above)
Comment (nategraf) [reply packages/contractkit/src/wrappers/Governance.ts]: Renamed this function and changed this signature to take a proposal.

**Pull request merged**